### DISCUSSION TITLE
Installation

### DISCUSSION THREAD

**Discussion by Yuanguo1994** [CsatiZoltan/CristalX#72] [Q&A] [answered]
Hi, CsatiZoltan

When I install the software, some errors occur, showing in the picture. Does it mean the installation has failed?
![image](https://user-images.githubusercontent.com/79046436/124108879-7afb4280-da5e-11eb-9ded-c44b2a148b63.png)

**Comment by CsatiZoltan** [comment, marked as answer]
Hi,

Only the installation of that specific [Python package](https://github.com/bluenote10/PandasDataFrameGUI) failed. The reason is that *git* is not installed on your system (as I see, you are working under Windows), therefore the command couldn't be executed. The *PandasDataFrameGUI* package is needed only for a small but useful functionality (to display grain statistics in a table). If you do not need that feature, you can leave it as it is. Otherwise, install *git* on your system. You can either install *git* only (e.g. through Anaconda: `conda install -c conda-forge git`), or by [installing a GUI](https://git-scm.com/download/gui/linux) (which will install *git* itself).

Next time, please send the full error message as text and not as a screenshot so that it is easier for me to debug.

Hope it helps,
Zoltan

P.S.:
To see what to expect from the dependency that you failed to install, here are two screenshots for a microstructure. Basically, it enables you to filter grains by certain properties (area, diameter, etc), and also display the frequency distribution of those properties.

![grains](https://user-images.githubusercontent.com/12742890/124117418-6b84f500-da70-11eb-8c55-c9ebd17356c1.png)
![histogram](https://user-images.githubusercontent.com/12742890/124117441-717ad600-da70-11eb-88aa-51809e7b5ee3.png)

**Reply by Yuanguo1994** [reply]
Thanks for your reply, it helps me solve the installation problem. However, when I was running the examples, another occurred, I attached the code and error message below:

skeleton=GS.create_skeleton(boundary)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-51-73e7d49657f8> in <module>
----> 1 skeletion= GS.create_skeleton(boundary)

C:\grains\segmentation.py in create_skeleton(self, boundary_image)
    211         skeleton = skeletonize(boundary_image)
    212         if self.__interactive_mode:
--> 213             io.imshow(skeleton)
    214             io.show()
    215             print('Skeleton constructed.')

~\Anaconda3\lib\site-packages\skimage\io\_io.py in imshow(arr, plugin, **plugin_args)
    157     if isinstance(arr, str):
    158         arr = call_plugin('imread', arr, plugin=plugin)
--> 159     return call_plugin('imshow', arr, plugin=plugin, **plugin_args)
    160 
    161 

~\Anaconda3\lib\site-packages\skimage\io\manage_plugins.py in call_plugin(kind, *args, **kwargs)
    205                                (plugin, kind))
    206 
--> 207     return func(*args, **kwargs)
    208 
    209 

~\Anaconda3\lib\site-packages\skimage\io\_plugins\matplotlib_plugin.py in imshow(image, ax, show_cbar, **kwargs)
    148     import matplotlib.pyplot as plt
    149 
--> 150     lo, hi, cmap = _get_display_range(image)
    151 
    152     kwargs.setdefault('interpolation', 'nearest')

~\Anaconda3\lib\site-packages\skimage\io\_plugins\matplotlib_plugin.py in _get_display_range(image)
     95         The name of the colormap to use.
     96     """
---> 97     ip = _get_image_properties(image)
     98     immin, immax = np.min(image), np.max(image)
     99     if ip.signed:

~\Anaconda3\lib\site-packages\skimage\io\_plugins\matplotlib_plugin.py in _get_image_properties(image)
     53                           (immin < lo or immax > hi))
     54     low_data_range = (immin != immax and
---> 55                       is_low_contrast(image))
     56     unsupported_dtype = image.dtype not in dtypes._supported_types
     57 

~\Anaconda3\lib\site-packages\skimage\exposure\exposure.py in is_low_contrast(image, fraction_threshold, lower_percentile, upper_percentile, method)
    649 
    650     dlimits = dtype_limits(image, clip_negative=False)
--> 651     limits = np.percentile(image, [lower_percentile, upper_percentile])
    652     ratio = (limits[1] - limits[0]) / (dlimits[1] - dlimits[0])
    653 

<__array_function__ internals> in percentile(*args, **kwargs)

~\Anaconda3\lib\site-packages\numpy\lib\function_base.py in percentile(a, q, axis, out, overwrite_input, interpolation, keepdims)
   3816     if not _quantile_is_valid(q):
   3817         raise ValueError("Percentiles must be in the range [0, 100]")
-> 3818     return _quantile_unchecked(
   3819         a, q, axis, out, overwrite_input, interpolation, keepdims)
   3820 

~\Anaconda3\lib\site-packages\numpy\lib\function_base.py in _quantile_unchecked(a, q, axis, out, overwrite_input, interpolation, keepdims)
   3935                         interpolation='linear', keepdims=False):
   3936     """Assumes that q is in [0, 1], and is an ndarray"""
-> 3937     r, k = _ureduce(a, func=_quantile_ureduce_func, q=q, axis=axis, out=out,
   3938                     overwrite_input=overwrite_input,
   3939                     interpolation=interpolation)

~\Anaconda3\lib\site-packages\numpy\lib\function_base.py in _ureduce(a, func, **kwargs)
   3513         keepdim = (1,) * a.ndim
   3514 
-> 3515     r = func(a, **kwargs)
   3516     return r, keepdim
   3517 

~\Anaconda3\lib\site-packages\numpy\lib\function_base.py in _quantile_ureduce_func(***failed resolving arguments***)
   4062         x_above = take(ap, indices_above, axis=0)
   4063 
-> 4064         r = _lerp(x_below, x_above, weights_above, out=out)
   4065 
   4066     # if any slice contained a nan, then all results on that slice are also nan

~\Anaconda3\lib\site-packages\numpy\lib\function_base.py in _lerp(a, b, t, out)
   3959 def _lerp(a, b, t, out=None):
   3960     """ Linearly interpolate from a to b by a factor of t """
-> 3961     diff_b_a = subtract(b, a)
   3962     # asanyarray is a stop-gap until gh-13105
   3963     lerp_interpolation = asanyarray(add(a, diff_b_a*t, out=out))

TypeError: numpy boolean subtract, the `-` operator, is not supported, use the bitwise_xor, the `^` operator, or the logical_xor function instead.

**Reply by CsatiZoltan** [reply]
Which example did you run? I cannot see it from the error traceback.

**Reply by Yuanguo1994** [reply]
"Which example did you run? I cannot see it from the error traceback."


The 'example_application' in the file called 'notebooks' in the software package. I searched the error in google, it is seemed caused by the update of python 3.8, causing the symbol '-' not supported anymore. But I failed in modifying the code

**Reply by CsatiZoltan** [reply]
The examples in the notebooks are meant to be run from within a Jupyter notebook. Now I realized that the conda environment does not contain Jupyter as a dependency. I will add to it soon. In the meantime, do the following:
1. Open the console and select the *CristalX* conda environment by typing `conda activate CristalX*.
2. Type `conda install jupyter` to install Jupyter into the *CristalX* environment.
3. Start a notebook by typing `jupyter notebook`. A web browser will show up.
4. Navigate to the directory where the file is (*CristalX/notebooks/*) and open *example_application.ipynb*.
5. You can now interactively explore the file.

I tried it on my machine and it works. Let me know if you have other problems.

**Reply by Yuanguo1994** [reply]
> The examples in the notebooks are meant to be run from within a Jupyter notebook. Now I realized that the conda environment does not contain Jupyter as a dependency. I will add to it soon. In the meantime, do the following:
> 
> 1. Open the console and select the _CristalX_ conda environment by typing `conda activate CristalX*.
> 2. Type `conda install jupyter` to install Jupyter into the _CristalX_ environment.
> 3. Start a notebook by typing `jupyter notebook`. A web browser will show up.
> 4. Navigate to the directory where the file is (_CristalX/notebooks/_) and open _example_application.ipynb_.
> 5. You can now interactively explore the file.
> 
> I tried it on my machine and it works. Let me know if you have other problems.

Thanks, the problem has been solved. Another error occurred in the step - 'image to geometry'. Similarly, I attached the command and error message below:


Command:
**from grains.cad import polygonize, search_neighbor, splinegonize
polygons = polygonize(extended_image, search_neighbor(2, np.inf), connectivity=1)
splinegons, _ = splinegonize(extended_image, search_neighbor(2, np.inf), connectivity=1, degree_min=3, degree_max=3, continuity='C0', tol=1)**

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-30-08f8398f17ee> in <module>
      1 from grains.cad import polygonize, search_neighbor, splinegonize
      2 polygons = polygonize(extended_image, search_neighbor(2, np.inf), connectivity=1)
----> 3 splinegons, _ = splinegonize(extended_image, search_neighbor(2, np.inf), connectivity=1, degree_min=3, degree_max=3, continuity='C0', tol=1)

~\CristalX\grains\cad.py in splinegonize(label_image, neighbor_search_algorithm, connectivity, detect_boundaries, degree_min, degree_max, continuity, tol)
    727         boundaries[region] = boundary
    728     # Handle the embedded grains
--> 729     splinegons = make_holes(splinegons)
    730     return splinegons, boundaries
    731 

~\CristalX\grains\cad.py in make_holes(splinegons)
    637     for i, Ai in splinegons.items():
    638         for j, Aj in splinegons.items():
--> 639             if i != j and np.isclose(face_area(subtract(Aj, Ai)), 0):  # Aj is a proper subset of Ai
    640                 splinegons[i] = subtract(splinegons[i], Aj)
    641     return splinegons

~\CristalX\grains\cad.py in subtract(shape1, shape2)
   1035     tools = TopTools_ListOfShape()
   1036     tools.Append(shape2)
-> 1037     difference = BRepAlgoAPI_Cut()
   1038     difference.SetTools(tools)
   1039     difference.SetArguments(arguments)

~\Anaconda3\envs\CristalX\lib\site-packages\OCC\BRepAlgoAPI.py in __init__(self, *args)
    335 
    336         """
--> 337         _BRepAlgoAPI.BRepAlgoAPI_Cut_swiginit(self, _BRepAlgoAPI.new_BRepAlgoAPI_Cut(*args))
    338 
    339     __repr__ = _dumps_object

TypeError: Wrong number or type of arguments for overloaded function 'new_BRepAlgoAPI_Cut'.
  Possible C/C++ prototypes are:
    BRepAlgoAPI_Cut::BRepAlgoAPI_Cut(TopoDS_Shape const &,TopoDS_Shape const &)
    BRepAlgoAPI_Cut::BRepAlgoAPI_Cut(TopoDS_Shape const &,TopoDS_Shape const &,BOPAlgo_PaveFiller const &,Standard_Boolean const)

**Reply by CsatiZoltan** [reply]
Yes, I know about this error. This is an error in the OpenCASCADE API. I will see what I can do.

**Reply by Yuanguo1994** [reply]
> Yes, I know about this error. This is an error in the OpenCASCADE API. I will see what I can do.



Is there any other way to get the splinegon surface? It seems in the example, you get the splinegon surface.

**Reply by CsatiZoltan** [reply]
The cause of the error is in version 0.17.3 of the [`pythonocc-core` package](https://github.com/CsatiZoltan/CristalX/blob/03f8b0f6cf2326b94787eebaa56d15de984ac3d2/environment.yml#L8). The reason I did not observe it before is that in my development environment, I installed `pythonocc-core` version 0.18.1 from sources, and that version works correctly. I don't recommend you to build `pythonocc-core` from sources as that would be complicated for Windows. Although version 0.18.x is available as a precompiled conda package, it has conflicts with some other packages that *CristalX* relies on. I have been planning for a long time (#19) to update `pythonocc` to a newer version, but that did not happen. I will do it soon because this functionality is a core feature of *CristalX*. In the meantime, as a quick fix, please comment out [this line](https://github.com/CsatiZoltan/CristalX/blob/03f8b0f6cf2326b94787eebaa56d15de984ac3d2/grains/cad.py#L729) in the *grains/cad.py* file. You will get the splinegon surfaces (I checked it). Actually, the `make_holes` method is only necessary if your grains form non-simply-connected domains.

**Reply by Yuanguo1994** [reply]
> The cause of the error is in version 0.17.3 of the [`pythonocc-core` package](https://github.com/CsatiZoltan/CristalX/blob/03f8b0f6cf2326b94787eebaa56d15de984ac3d2/environment.yml#L8). The reason I did not observe it before is that in my development environment, I installed `pythonocc-core` version 0.18.1 from sources, and that version works correctly. I don't recommend you to build `pythonocc-core` from sources as that would be complicated for Windows. Although version 0.18.x is available as a precompiled conda package, it has conflicts with some other packages that _CristalX_ relies on. I have been planning for a long time (#19) to update `pythonocc` to a newer version, but that did not happen. I will do it soon because this functionality is a core feature of _CristalX_. In the meantime, as a quick fix, please comment out [this line](https://github.com/CsatiZoltan/CristalX/blob/03f8b0f6cf2326b94787eebaa56d15de984ac3d2/grains/cad.py#L729) in the _grains/cad.py_ file. You will get the splinegon surfaces (I checked it). Actually, the `make_holes` method is only necessary if your grains form non-simply-connected domains.

Thanks for your answer, now I can run through the whole example, and move to my microstructure segmentation. Really thanks for this, it helps me a lot!!!!

**Reply by CsatiZoltan** [reply]
You are welcome. Feel free to open a new discussion thread if you face other problems.
